mqlc/mqlc.go

```diff
@@ -1684,18 +1684,111 @@ func (c *compiler) postCompile() {
 			chunk, typ, ref := c.expandListResource(chunk, ref)
 			switch chunk.Id {
 			case "$one", "$all", "$none", "$any":
+				// default fields
 				ref = chunk.Function.Binding
-				chunk := code.Chunk(chunk.Function.Binding)
+				chunk := code.Chunk(ref)
 				typ = types.Type(chunk.Function.Type)
-				c.expandResourceFields(chunk, typ, ref)
-				block.Datapoints = append(block.Datapoints, block.TailRef(ref))
+				expanded := c.expandResourceFields(chunk, typ, ref)
+				// when no defaults are definied or query isn't about a resource, no block was added
+				if expanded {
+					block.Datapoints = append(block.Datapoints, block.TailRef(ref))
+					c.addValueFieldChunks(ref)
+				}
 			default:
 				c.expandResourceFields(chunk, typ, ref)
 			}
 		}
 	}
 }
 
+// addValueFieldChunks takes the value fields of the assessment and adds them to the
+// block forthe default fields
+// This way, the actual data of the assessment automatically shows up in the output
+// of the assessment that failed the assessment
+func (c *compiler) addValueFieldChunks(ref uint64) {
+	var assessmentBlock *llx.Block
+	var whereChunk *llx.Chunk
+
+	// find chunk with where/whereNot function
+	for {
+		chunk := c.Result.CodeV2.Chunk(ref)
+		if chunk.Function == nil {
+			// this is a safe guard and shouldn't happen
+			log.Error().Msg("failed to find where function for assessment")
+			return
+		}
+		if chunk.Function.Args != nil {
+			whereChunk = chunk
+			break
+		}
+		ref = chunk.Function.Binding
+	}
+
+	// find the referenced blcok for the where function
+	for i := len(whereChunk.Function.Args) - 1; i >= 0; i-- {
+		arg := whereChunk.Function.Args[i]
+		if types.Type(arg.Type).Underlying() == types.FunctionLike {
+			raw := arg.RawData()
+			blockRef := raw.Value.(uint64)
+			assessmentBlock = c.Result.CodeV2.Block(blockRef)
+			break
+		}
+	}
+
+	defaultFieldsBlock := c.Result.CodeV2.Blocks[len(c.Result.CodeV2.Blocks)-1]
+	defaultFieldsRef := defaultFieldsBlock.Chunks[1].Function.Binding
```

Comment:
isn't this the first chunk in the block `defaultFieldsBlock | 1`?

Reply:
It's the reference to the first chunk. I didn't find a way to get the "address" of a chunk. I only found it in the binding.

Reply:
I think it gets done by `blockRef << 32 | 1`. You'll see that a lot in the tests. Probably should be a function to do that

Reply:
Ah, thanks.

I added some functions and changed the above code accordingly.